// raptorjuice/Gruntfile.js
module.exports = function(grunt) {

  grunt.initConfig({

    pkg : grunt.file.readJSON('package.json'), 
    
    jsdoc : {
      dist : {
        src : [ 'index.js', 'README.md', 'LICENSE' ],
        options : {
          destination : 'doc',
          template : "node_modules/ink-docstrap/template/",
          configure : "jsdoc.json"
        }
      }
    },

    clean : {
      doc : 'doc'
    },

    mochaTest : {
      options : {
        reporter : 'spec'
      },
      test : {
        src : 'spec/**/*.spec.js'
      }
    },

    browserify : {
      dist : {
        files : {
          './dist/raptor.js' : './index.js'
        }
      }
    },

    uglify : {
      dist : {
        files : {
          './dist/raptor.min.js' : './dist/raptor.js'
        }
      }
    },

    watch : {
      all : {
        options : {
          atBegin : true,
        },
        files : [ 'index.js', 'spec/**/*.spec.js' ],
        tasks : [
          'mochaTest', 'browserify', 'uglify', 'clean', 'jsdoc:dist'
        ]
      }
    }

  });

  grunt.loadNpmTasks('grunt-browserify');
  grunt.loadNpmTasks('grunt-jsdoc');
  grunt.loadNpmTasks('grunt-mocha-test');
  grunt.loadNpmTasks('grunt-contrib-clean');
  grunt.loadNpmTasks('grunt-contrib-watch');
  grunt.loadNpmTasks('grunt-contrib-uglify');

  grunt.registerTask('test', 'mochaTest');

};
